<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=
    , initial-scale=1.0">
    <title>Document</title>

    <script type="text/javascript" src="js/script.js"></script>
    <style>
        div {
            width: 200px;
            height: 200px;
            background: #58a;
        }
    </style>

</head>

<body>

    <!-- Promise基本语法 -->
    <script>
        const p = new Promise(function (resolve, reject) {
            setTimeout(() => {
                let data = 'i am resolve'
                resolve(data)

                // let err = 'i am reject'
                // reject(err)
            }, 1000);
        })

        // p.then(function (value) {
        //     console.log(value)
        // }, function (reason) {
        //     console.log(reason)
        // })

    </script>


    <!-- ajax -->
    <script>
        // 接口地址: https://api.apiopen.top/getJoke
        const p3 = new Promise((resolve, reject) => {
            //1. 创建对象
            const xhr = new XMLHttpRequest();

            //2. 初始化
            xhr.open("GET", "https://api.apiopen.top/getJ");

            //3. 发送
            xhr.send();

            //4. 绑定事件, 处理响应结果
            xhr.onreadystatechange = function () {
                //判断
                if (xhr.readyState === 4) {
                    //判断响应状态码 200-299
                    if (xhr.status >= 200 && xhr.status < 300) {
                        //表示成功
                        resolve(xhr.response);
                    } else {
                        //如果失败
                        reject(xhr.status);
                    }
                }
            }
        })

        //指定回调
        p3.then(function (value) {
            console.log(value);
        }, function (reason) {
            console.error(reason);
        });
    </script>


    <!-- then, catch -->
    <script>

        //创建 promise 对象
        const p4 = new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve('用户数据');
                // reject('出错啦');
            }, 1000)
        });

        //调用 then 方法  then方法的返回结果是 Promise 对象, 对象状态由回调函数的执行结果决定
        //1. 如果回调函数中返回的结果是 非promise 类型的属性, 状态为成功, 返回值为对象的成功的值
        const result = p4.then(value => {
            console.log(value);
            //1. 非 promise 类型的属性
            // return 'iloveyou';
            //2. 是 promise 对象
            // return new Promise((resolve, reject)=>{
            //     // resolve('ok');
            //     reject('error');
            // });
            //3. 抛出错误
            // throw new Error('出错啦!');
            throw '出错啦!';
        }, reason => {
            console.warn(reason);
        });

        //链式调用
        p4.then(value => {
        }).then(value => {
        });

        //catch
        p.catch(function(reason){
            console.warn(reason);
        });

    </script>





</body>

</html>